modifiers/control_signals.cpp

```diff
@@ -0,0 +1,102 @@
+/*********************                                                        */
+/*! \file control_signals.cpp
+** \verbatim
+** Top contributors (to current version):
+**   Makai Mann
+** This file is part of the pono project.
+** Copyright (c) 2019 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief Useful functions for adding control signal semantics to a system.
+**        For example, toggling a clock or setting a reset sequence.
+**
+**/
+
+#include "assert.h"
+#include "math.h"
+
+#include "modifiers/control_signals.h"
+
+using namespace smt;
+using namespace std;
+
+namespace pono {
+
+void toggle_clock(TransitionSystem & ts, const Term & clock_symbol)
+{
+  const SmtSolver & s = ts.solver();
+  Sort sort = clock_symbol->get_sort();
+  SortKind sk = sort->get_sort_kind();
+  Sort one_bit_sort = s->make_sort(BV, 1);
+
+  if (clock_symbol->get_sort() != one_bit_sort && sk != BOOL) {
+    throw PonoException("Expecting a boolean or one-bit clock sort.");
+  }
+
+  Term zero = s->make_term(0, one_bit_sort);
+  Term clk_state = clock_symbol;
+  assert(!ts.is_next_var(clk_state));
+  if (!ts.is_curr_var(clk_state)) {
+    Term clk_state = ts.make_statevar(clock_symbol->to_string() + "__state__",
+                                      clock_symbol->get_sort());
+    ts.constrain_inputs(s->make_term(Equal, clock_symbol, clk_state));
+  }
+  assert(sk == BV || sk == BOOL);
+  if (sk == BV) {
+    ts.constrain_init(s->make_term(Equal, clk_state, zero));
+    ts.assign_next(clk_state, s->make_term(BVNot, clk_state));
+  } else if (sk == BOOL) {
+    ts.constrain_init(s->make_term(Not, clk_state));
+    ts.assign_next(clk_state, s->make_term(Not, clk_state));
```

Comment:
I am not sure how realistic the following scenario is, but what happens if e.g., a BTOR2 file already contains logic to handle a custom clock signal? In that case, the user would not need to use the new add-clock feature, however, it would be good to have some error checking to prevent mixing an existing clock logic with the clock logic we add to the model.  

In the example below, a counter is incremented whenever the clock (already part of the model) is high, and the clock is always kept high by its next-function. Hence the property fails, i.e., the counter reaches its maximum value.

The call `./pono count2-clk.btor2` produces a counterexample as expected.

However, the call `./pono -k 100 -c clk count2-clk.btor2` reports `unknown`. We could throw an exception here because the clock signal gets two next-functions assigned, which we usually prevent, if I am not mistaken. Also, with a toggling clock that we add, the property should be reached in twice the number of steps, but it isn't.
 
```
1 sort bitvec 3
2 zero 1
3 state 1 count
4 init 1 3 2
5 one 1
6 add 1 3 5
8 ones 1
9 sort bitvec 1
10 eq 9 3 8
11 bad 10
12 zero 9
; clock signal
13 state 9 clk
; clock initialized to zero
14 init 9 13 12
15 ite 1 13 6 3
; next(count): if clk is high then count++ else don't change
16 next 1 3 15
; next-func of clk -- for testing, keep clk always high
17 one 9
18 next 9 13 17
```